### ISSUE TITLE
Refactor module configuration

### ISSUE THREAD

**Issue by lfcnassif** [sepinf-inc/IPED#538]
To make easier the implementation of #23 and #24, the configuration loading/saving will need to be refactored, using some previous work by @patrickdalla and some ideas of @atilaromero (https://github.com/sepinf-inc/IPED/issues/23#issuecomment-596144210)

**Comment by tc-wleite** [comment]
@lfcnassif, not sure how related to what you are working with in this issue, but a minor suggestion:
Couldn't the retrieval of configurations, use in many places in the code, be simpler, as always the first (and only) object is retrieved?

```
//Something like...
IPEDConfig ipedConfig = (IPEDConfig) ConfigurationManager.findObject(IPEDConfig.class); 

//...instead of
IPEDConfig ipedConfig = (IPEDConfig) ConfigurationManager.getInstance().findObjects(IPEDConfig.class).iterator().next(); 
```

**Comment by lfcnassif** [comment]
For sure! I plan to simplify it, it is contributed code.

**Comment by lfcnassif** [comment]
Do you know if it is possible to avoid  the cast? I tried some things, but didn't manage to make it work

**Comment by lfcnassif** [comment]
Another question (a lot will come), should a Task access just its configuration or may it access all of them? Init() signatute change depends on this decision.

**Comment by tc-wleite** [comment]
> Do you know if it is possible to avoid the cast? I tried some things, but didn't manage to make it work

```
//Something like this? 
IPEDConfig ipedConfig = ConfigurationManager.findObject(IPEDConfig.class);

//In ConfigurationManager...
public static <T extends Configurable> T findObject(Class<? extends Configurable> clazz) {
    for (Configurable configurable : singleton.loadedConfigurables.keySet()) {
        if (configurable.getClass().equals(clazz)) {
            return (T)configurable;
        }
    }
    return null;
}
```

**Comment by lfcnassif** [comment]
I am thinking about a new abstract method for Tasks to return a Configurable object, so new devs will know they could create one, and a future UI could ask Tasks what config they use. Still elaborating the idea... What do you think?

**Comment by tc-wleite** [comment]
> Another question (a lot will come), should a Task access just its configuration or may it access all of them? Init() signature change depends on this decision.

Not sure. I think currently there are cases of such "dependencies" (tasks accessing other tasks configuration), right?

**Comment by tc-wleite** [comment]
> I am thinking about a new abstract method for Tasks to return a Configurable object, so new devs will know they could create one, and a future UI could ask Tasks what config they use. Still elaborating the idea... What do you think?

Seems a good idea!

**Comment by lfcnassif** [comment]
> ```
> //Something like this? 
> IPEDConfig ipedConfig = ConfigurationManager.findObject(IPEDConfig.class);
> 
> //I *think* this works... In ConfigurationManager:
> public static <T extends Configurable> T findObject(Class<? extends Configurable> clazz) {
>     for (Configurable configurable : singleton.loadedConfigurables.keySet()) {
>         if (configurable.getClass().equals(clazz)) {
>             return (T)configurable;
>         }
>     }
>     return null;
> }
> ```

Perfect! I'm still not very used with generics syntax...

**Comment by lfcnassif** [comment]
> Not sure. I think currently there are cases of such "dependencies" (tasks accessing other tasks configuration), right?

Yes. One Configurable could load more than one file, but maybe it is not a good idea to one file being read by more than one Configurable...

**Comment by lfcnassif** [comment]
> Perfect! I'm still not very used with generics syntax...

E.g. why the 'T' is needed before 'findObject'... but don't waste your time answering me, I must study the doc...

**Comment by tc-wleite** [comment]
> > Not sure. I think currently there are cases of such "dependencies" (tasks accessing other tasks configuration), right?
> 
> Yes. One Configurable could load more than one file, but maybe it is not a good idea to one file being read by more than one Configurable, could make maintenance more difficult...

Indeed! If it is possible to limit to a single one it would be better, but I don't know if that would require other workarounds in existing tasks.

**Comment by lfcnassif** [comment]
For those learning like me, T is the normal return type declaration, <T extends Configurable> is the new type declaration, not the opposite:
https://docs.oracle.com/javase/tutorial/java/generics/methods.html

**Comment by lfcnassif** [comment]
hey @tc-wleite I started to implement in all tasks a new `List<Configurable> getConfigurables()` method and was going to change init(...) to receive just those configurables, I thought it would be more general to handle more cases. But stopped when thinking about configuration UI, my original idea was to show 'one' configuration dialog for each task. Maybe tasks returning more than one Configurable could open a dialog with multiple tabs (e.g. ExportFileTask would show one tab for categories to export and other tab for keywords to look for). What do you think? (CategoriesToExport.txt and KeywordsToExport.txt could be also merged...)

And maybe IPEDConfig could be always passed to init(List<Configurables>) similar to current first parameter. Tomorrow I will try to take a closer look at all tasks to find possible duplicated parameters that would look strange if shown in different config dialogs. I "think" dependencies between tasks would not be a big issue, but not sure where to check the dependencies, in UI or in tasks code somehow... maybe another method to check other loaded Configurables or to return needed Tasks classes to check isEnabled() (just thought about that)... I would like to not add too much complexity for new devs... any suggestion is welcome.

**Comment by lfcnassif** [comment]
> Indeed! If it is possible to limit to a single one it would be better, but I don't know if that would require other workarounds in existing tasks.

What did you mean here? To limit to one configurable per task? Or to limit loading one specific config file to one configurable?

**Comment by tc-wleite** [comment]
> > Indeed! If it is possible to limit to a single one it would be better, but I don't know if that would require other workarounds in existing tasks.
> 
> What did you mean here? To limit to one configurable per task? Or to limit loading one specific config file to one configurable?

I was thinking about the former: one Configurable <--> one Task.
"General" Configurables (LocalConfig and IPEDConfig) would have be available too, right?

**Comment by tc-wleite** [comment]
> I was thinking about the former: one Configurable <--> one Task.
> "General" Configurables (LocalConfig and IPEDConfig) would have be available too, right?

On a second thought, not sure if this makes sense.
For example, a configuration that is in IPEDConfig today (like if a task should be enabled), would it be accessed by the task in its own Configurable? Or in a general Configurable object?
The first option seems good thinking about the task implementation, but would the UI to view/edit parameters keep the current organization (all tasks enable/disable parameter in a single place)?
Not sure how you are dealing with that.

**Comment by lfcnassif** [comment]
> I was thinking about the former: one Configurable <--> one Task.
> "General" Configurables (LocalConfig and IPEDConfig) would have be available too, right?

I just thought about IPEDConfig, but LocalConfig could be made available too if needed.

**Comment by lfcnassif** [comment]
> For example, a configuration that is in IPEDConfig today (like if a task should be enabled), would it be accessed by the task in its own Configurable? Or in a general Configurable object?
That's my doubt.

> The first option seems good thinking about the task implementation, but would the UI to view/edit parameters keep the current organization (all tasks enable/disable parameter in a single place)?
> Not sure how you are dealing with that.
I think current organization could be kept to easy migration of old users, just have to think about multiple tasks configurables writing to the same file (IPEDConfig.txt) when being enabled/disabled. That would not be needed if we use serialized configurables to start processing, but would be useful to save new profiles...

**Comment by lfcnassif** [comment]
Another idea is to get rid of IPEDConfig and enable/disable tasks in TaskInstaller.xml, the UI will load tasks from there (like processing engine does) and that xml could have an 'enable' param for each task, so tasks won't need to know if they are enabled or not, they won't be loaded... Possibly users will not like that...

I think about moving some IPEDConfig options to module options (excludeKnown would be a config of HashDBLookup), hashes would be a config of HashTask, indexUnknown/Corrupted/Unallocated an option of IndexTask, enableFileParsing and expandContainers could be options of ParsingTask...

I still don't known about options "shared" by more than one module like addUnallocated (affects CarvingTask and IndexTask), actually it is an option of SleuthkitDatasourceReader...

Or maybe the interface could list all configurables, independently if they are related to tasks or not...

More thoughts than certainty...

**Comment by lfcnassif** [comment]
maybe @bhoelz could give some recommendation here?

**Comment by tc-wleite** [comment]
> Another idea is to get rid of IPEDConfig and enable/disable tasks in TaskInstaller.xml, the UI will load tasks from there (like processing engine does) and that xml could have an 'enable' param for each task, so tasks won't need to know if they are enabled or not, they won't be loaded... Possibly users will not like that...

Controlling enabled taks in TaskInstaller.xml seems a good idea thinking about the code/project organization, but as I said, for the users, the current way (IPEDConfig.txt) seems a bit easier. 

> I think about moving some IPEDConfig options to module options (excludeKnown would be a config of HashDBLookup), hashes would be a config of HashTask, indexUnknown/Corrupted/Unallocated an option of IndexTask, enableFileParsing and expandContainers could be options of ParsingTask...

Moving these options from IPEDConfig (and maybe some others from AdvancedConfig, not sure) to task specific Configurables seems a great idea.
 
> I still don't known about options "shared" by more than one module like addUnallocated (affects CarvingTask and IndexTask), actually it is an option of SleuthkitDatasourceReader...
> Or maybe the interface could list all configurables, independently if they are related to tasks or not...

These "other" options (not related to tasks or shared by more than one tasks) could go to other Configurables, not direct related to tasks. Tasks could use them, but the UI would need to list them using another identification (not related to tasks). 
If these other options can be grouped, it would be nice, but probably few of them would need to a go to a generic Configurable (similar to the current AdvancedConfig plus parts of IPEDConfig).

**Comment by lfcnassif** [comment]
Thanks @tc-wleite! I think we are converging, also thought about moving some AdvancedConfig options, I agree on all points. I will continue to draft the proposal, thanks.

**Comment by lfcnassif** [comment]
I created some classes to help common configurations:
1. AbstractTaskConfig: loads enable param from IPEDConfig and process a task specific config file
2. AbstractTaskPropertiesConfig: same as before, but config file is a property file
3. EnableTaskProperty: simple configurable impl for tasks that just have enable param in IPEDConfig

I think configurables are easier to implement now.
And possibly I will change List<Configurable> AbstractTask.getConfigurables() to return a single configurable as you suggested. For now, just ExportFileTask returns multiple configurables, I think about creating some CompositeConfigurable keeping an internal list of Configurables, this should take care of exceptional cases like ExportFileTask.

What do you think @tc-wleite?

**Comment by lfcnassif** [comment]
Thanks @tc-wleite, any other suggestion is welcome. Planned next steps:
1. Convert remaining Tasks to use new approach
2. Change init() signature to receive Configurable(s) or ConfigurationManager, not sure which is better
3. Move parameters of IPEDConfig and AdvancedConfig to new Tasks configs
4. Create Configurable write/save abstract method and impls

**Comment by lfcnassif** [comment]
About "ignoreHardLinks" option currently in AdvancedConfig.txt. I moved it to IPEDConfig.txt, because it is implemented as a Task today. But I think maybe it makes more sense to put it in a more generic FileSystemConfig.txt together with addUnallocated, addFileSlacks, minOrphanSizeToIgnore, unallocatedFragSize... Any thoughts?

**Comment by lfcnassif** [comment]
> 2\. Change init() signature to receive the task specific List<Configurable> or the ConfigurationManager instance, not sure which is better

@tc-wleite what do you think about this? init() could be also left no-arg and Configurables can be accessed statically from ConfigurationManager.

**Comment by tc-wleite** [comment]
> @tc-wleite what do you think about this? init() could be also left no-arg and Configurables can be accessed statically from ConfigurationManager.

I slightly prefer receiving "something" (also not sure if the manager or a task specific list) in init(), instead of statically access the ConfigurationManager. It seems a bit clearer for new developers.
But I guess this first option would require more work to change all existing code, right?

**Comment by lfcnassif** [comment]
> It seems a bit clearer for new developers.

I agree with this.

> But I guess this first option would require more work to change all existing code, right?

Yes, but not that work. I will finish the implementation in remaining tasks and see what fits better, thanks!

**Comment by lfcnassif** [comment]
So, I think I finished the configs loading refactoring. I was going to implement configs saving, but I think it is better to merge this as it is now to avoid future conflicts with other PRs like #593. Relevant changes to IPEDConfig.txt:
1. indexUnallocated was moved to conf/IndexTaskConfig.txt
2. indexUnknown and indexCorrupted were moved to conf/ParsingTaskConfig.txt (and renamed to parseUnknown and parseCorrupted)
3. addUnallocated and addFileSlacks were moved to conf/FileSystemConfig.txt
4. excludeKnown was moved to conf/HashDBLookupConfig.txt
5. enableOCR was moved to conf/OCRConfig.txt (in UI will be shown as a subconfig of ParsingTask, it is not an independent task)

I plan to write a migration guide to version 4.0 with those and other things to make users life easier. I think the new grouping makes sense, but some (or all) of those changes could be reverted, the logical grouping in UI will be independent of grouping in config files. What other devs think about new organization? Should I revert some changes?

**Comment by lfcnassif** [comment]
@tc-wleite @fmpfeifer @hauck-jvsh @streeg

**Comment by hauck-jvsh** [comment]
This is a huge refactor, the indexUnknow and indexCorrupted only parses raw strings? If so, as you are already changing the name, I think that it should be more suggestive. As for the others I think they are good changes. As you suggest user's will need a good documentation to follow, there are several breaking changes, not only in this part (Ex. unification of all languages profiles). Another question, the UI will be mandatory? I think there are users who will still prefer to use only by command line.

**Comment by lfcnassif** [comment]
Thanks @hauck-jvsh.

> the indexUnknow and indexCorrupted only parses raw strings?

Yes. Actually new names are parseUnknownFiles and parseCorruptedFiles. Some possible names: parse(Raw|Plain|Printable)StringsFrom(Unknown|Corrupted)Files? 

> Another question, the UI will be mandatory?

No it will be optional, but it will easy a lot the usage of #24

**Comment by hauck-jvsh** [comment]
Raw|Plain|Printable are different parsers used in those files? If so I think parse(Unknown|Corrupted)Files may be a list of which parsers should be executed for each one, for example: parseUnknownFiles= Raw, Printable. If those were just name suggestions for the same parser I would prefer parseRawStringsFrom(Unknown|Corrupted)Files. But it is just my opinion, it would be better to have more ideas.

**Comment by lfcnassif** [comment]
> Raw|Plain|Printable are different parsers used in those files?

just different name suggestions for the same parser.

**Comment by tc-wleite** [comment]
> I plan to write a migration guide to version 4.0 with those and other things to make users life easier. I think the new grouping makes sense, but some (or all) of those changes could be reverted, the logical grouping in UI will be independent of grouping in config files. What other devs think about new organization? Should I revert some changes?

I like the new organization. 
So far I don't have anything to add.

**Comment by lfcnassif** [comment]
Reopening, still need to implement config saving.

**Comment by lfcnassif** [comment]
> still need to implement config saving.

I was going to implement config saving to original/alternative txt/xml/json files, but, could the UI just serialize all configs to a single file? This would save me a lot of work... Of course the single serialized config could be opened by UI for review or used by cmd line processing later.

**Comment by tc-wleite** [comment]
@lfcnassif, I know you are still working on this, but I found a minor issue in RegexTask (NPE during processing).
It seems that the class variable `enabled` (used in `process()`) and `regexConfig.isEnabled()` (used in `init()`) don't have the same value.

**Comment by lfcnassif** [comment]
Thanks @tc-wleite, will take a look. Tests are always welcome!

**Comment by lfcnassif** [comment]
just fixed in master, thanks!

**Comment by lfcnassif** [comment]
Have you disabled it because of slow init time after #558? I'm considering to comment out the pattern for pt-BR seeds and maybe leave just english pattern on. What do you think? But possibly it is more annoying for devs...

**Comment by tc-wleite** [comment]
> Have you disabled it because of slow init time after #558? I'm considering to comment out the pattern for pt-BR seeds and maybe leave just english pattern on. What do you think? But possibly it is more annoying for devs...

No, I was testing a new feature and disabled most of the tasks. 
I did notice that after #558 RegexTask requires a bit longer initialization, but I found it perfectly acceptable.

**Comment by lfcnassif** [comment]
I "think" this was finished. We can reopen if needed.

**Comment by lfcnassif** [comment]
Reopening, TempFileTask is running on fastmode profile, causing slowness.

**Comment by lfcnassif** [comment]
> Reopening, TempFileTask is running on fastmode profile, causing slowness.

Wrong issue, closing.

**Comment by lfcnassif** [comment]
All devs,

We are almost finishing the 4.0.0 roadmap. Should any config file changes be rolled back (e.g. move `enableOCR` back to IPEDConfig.txt although it's really a ParsingTask config and not another independent module)? Should `enableMinIO` and `enableElasticSearch` be moved to IPEDConfig to follow current pattern of enabling/disabling modules in that main config file? Some changed configs below:

> * indexUnallocated was moved to conf/IndexTaskConfig.txt
> * indexUnknown and indexCorrupted were moved to conf/ParsingTaskConfig.txt (and renamed to parseUnknown and parseCorrupted)
> * addUnallocated and addFileSlacks were moved to conf/FileSystemConfig.txt
> * excludeKnown was moved to conf/HashDBLookupConfig.txt
> * enableOCR was moved to conf/OCRConfig.txt (in UI will be shown as a subconfig of ParsingTask, it is not an independent task)

**Comment by hauck-jvsh** [comment]
As all other modules are enabled/disabled in the IPEDConfig, I think that  elastic and MinIO modules should be there too.

**Comment by lfcnassif** [comment]
Reopening to move Elastic/MinIO enable options to main config file.
